File: examples/android/app/src/main/java/com/jnibridge/modules/dom/yoga/YogaUINode.java
package com.jnibridge.modules.dom.yoga;

import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ScrollView;

import com.facebook.yoga.YogaConstants;
import com.facebook.yoga.YogaDisplay;
import com.facebook.yoga.YogaFlexDirection;
import com.facebook.yoga.YogaNode;
import com.facebook.yoga.YogaNodeFactory;
import com.jnibridge.EngineScope;
import com.jnibridge.modules.dom.IUINode;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public abstract class YogaUINode implements IUINode {
    protected YogaNode _yogaNode;
    protected String _tagName;
    protected int _uniqueId;
    protected EngineScope _scope;
    protected View _view;
    public YogaUINode(String tagName, int uniqueId, EngineScope scope) {
        this(tagName, uniqueId, scope, YogaNodeFactory.create());
    }

    public YogaUINode(String tagName, int uniqueId, EngineScope scope, YogaNode node) {
        _tagName = tagName;
        _uniqueId = uniqueId;
        _scope = scope;
        _yogaNode = node == null ? YogaNodeFactory.create() : node;
        getYogaNode().setDisplay(YogaDisplay.FLEX);
        getYogaNode().setFlexDirection(YogaFlexDirection.COLUMN);
    }

    public void setWidth(float width) {
        _yogaNode.setWidth(width);
    }

    public void setHeight(float height) {
        _yogaNode.setHeight(height);
    }

    public void setBackgroundColor(int color) {
        getNativeView().setBackgroundColor(color);
    }

    public YogaNode getYogaNode() {
        return _yogaNode;
    }

    public void setYogaNode(YogaNode node) {
        _yogaNode = node;
    }

    public View getNativeView() {
        return _view;
    }

    protected void setNativeView(View view) {
        _view = view;
        _view.setTag(this);
        getYogaNode().setData(_view);
        getNativeView().setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Log.d("JSEngine","click tagName=" + getTagName());
                fireUIEvent("click");
            }
        });
        getNativeView().setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        fireUIEvent("touchstart");
                    case MotionEvent.ACTION_MOVE:
                        fireUIEvent("touchmove");
                    case MotionEvent.ACTION_UP:
                        fireUIEvent("touchend");
                    case MotionEvent.ACTION_CANCEL:
                        fireUIEvent("touchcancel");
                    case MotionEvent.ACTION_OUTSIDE:
                        fireUIEvent("touchoutside");
                }
                return false;
            }
        });
    }

    public void calculateLayout() {
        getYogaNode().calculateLayout(YogaConstants.UNDEFINED, YogaConstants.UNDEFINED);
    }

    @Override
    public EngineScope getScope() {
        return _scope;
    }

    @Override
    public int getUniqueId() {
        return _uniqueId;
    }

    @Override
    public String getTagName() {
        return _tagName;
    }

    @Override
    public ArrayList<IUINode> getChildNodes() {
        return null;
    }

    @Override
    public String getAttribute(String key) {
        String value = "";
        switch (key) {
            case "width":
                calculateLayout();
                value = String.valueOf(YogaStyleHelper.toDisplayUnit(getYogaNode().getLayoutWidth()));
                break;
            case "height":
                calculateLayout();
                value = String.valueOf(YogaStyleHelper.toDisplayUnit(getYogaNode().getLayoutHeight()));
                break;
        }
        Log.d("JSEngine", "getAttribute key=" + key + ", value=" + value);
        return value;
    }

    @Override
    public void setAttribute(String key, String value) {
        switch (key) {
            case "width":
            case "height":
                this.setStyle(key, value);
                break;
        }
    }

    @Override
    public void setStyle(String key, String value) {
        YogaStyleHelper.setStyle(getYogaNode(), key, value);
    }

    @Override
    public void resetStyle() {
        YogaStyleHelper.resetStyle(getYogaNode());
    }

    @Override
    public void removeAttribute(String key) {

    }

    @Override
    public IUINode appendChild(IUINode newChild) {
        return null;
    }

    @Override
    public boolean contains(IUINode child) {
        return false;
    }

    @Override
    public IUINode insertBefore(IUINode newChild, IUINode refChild) {
        return null;
    }

    @Override
    public IUINode removeChild(IUINode child) {
        return null;
    }

    @Override
    public IUINode replaceChild(IUINode newChild, IUINode oldChild) {
        return null;
    }

    public void fireUIEvent(String eventType) {
        fireUIEvent(eventType, null);
    }

    public void fireUIEvent(String eventType, JSONObject obj) {
        JSONObject json = new JSONObject();
        try {
            json.put("uniqueId", _uniqueId);
            json.put("type", eventType);
            if(obj != null) {
                json.put("data", obj);
            }
        } catch (JSONException jsonException) {
            jsonException.printStackTrace();
        }
        getScope().invokeJSModule("JSMUIEvent", "onNativeEvent", json.toString());
    }
}

File: examples/ios/ioscplus/ioscplus/YogaUINode.swift
//
//  YogaUINode.swift
//  ioscplus
//
//  Created by song on 2021/8/24.
//

import Foundation
import UIKit
import YogaKit
import UIColor_Hex_Swift
public class YogaUINode : IUINode {
    private var _view: UIView! = nil
    private var _userStyle: [String:String] = [:]
    public override init(tagName: String, uniqueId: Int32, scope: JSEngineScope) {
        super.init(tagName: tagName, uniqueId: uniqueId, scope: scope)
    }

    public func getUserStyle(key: String) -> String? {
        return _userStyle[key];
    }
    
    public func setUserStyle(key: String, value: String) {
        _userStyle.updateValue(value, forKey: key)
    }
    
    public func getNativeView() -> UIView {
        return _view;
    }
    
    public func setNativeView(view: UIView) {
        view.yoga.isEnabled = true
        view.yoga.flexDirection = .column
        view.tag = Int(getUniqueId())
        _view = view;
    }
    
    public func getYogaLayout() -> YGLayout {
        return _view.yoga;
    }
    
    public override func getAttribute(key: String) -> String {
        var value = "";
        switch (key) {
            case "width":
                let width = getNativeView().bounds.width
                value = "\(width)"
                break;
            case "height":
                let height = getNativeView().bounds.height
                value = "\(height)"
                break;
            default:
                value = super.getAttribute(key: key)
            break;
        }
        return value
    }
    
    public override func setAttribute(key: String, value: String) {
        switch (key) {
            case "width":
                setStyle(key: key, value: value)
                break;
            case "height":
                setStyle(key: key, value: value)
                break;
            default:
                super.setAttribute(key: key, value: value)
            break;
        }
    }
    
    public override func setStyle(key: String, value: String) {
        setUserStyle(key: key, value: value)
        switch (key) {
            case "backgroundColor":
                getNativeView().backgroundColor = UIColor(value);
                break;
            default:
                YogaStyleHelper.setStyle(node: self, key: key, value: value)
            break;
        }
    }
    
    public override func resetStyle() {
        YogaStyleHelper.resetStyle(node: self, userStyle: _userStyle);
    }
}
